src/main/java/com/conveyal/analysis/controllers/RegionalAnalysisController.java

```diff
@@ -156,18 +168,181 @@ private int getIntQueryParameter (Request req, String parameterName, int default
+                    break;
+                case PNG:
+                    grid.writePng(fos);
+                    break;
+                case GEOTIFF:
+                    grid.writeGeotiff(fos);
+                    break;
+            }
+
+            fileStorage.moveIntoStorage(singleCutoffFileStorageKey, localFile);
+        }
+        return singleCutoffFileStorageKey;
+    }
+
+    private Object getAllRegionalResults (Request req, Response res) throws IOException {
+        // Get the UUID of the regional analysis for which we want the output data.
+        final String regionalAnalysisId = req.params("_id");
+        final UserPermissions userPermissions = UserPermissions.from(req);
+        RegionalAnalysis analysis = Persistence.regionalAnalyses.findPermitted(
+                QueryBuilder.start("_id").is(req.params("_id")).get(),
+                DBProjection.exclude("request.scenario.modifications"),
+                userPermissions
+        ).iterator().next();
+        if (analysis == null || analysis.deleted) {
+            throw AnalysisServerException.notFound("The specified regional analysis is unknown or has been deleted.");
+        }
+        FileStorageKey zippedResultsKey = new FileStorageKey(RESULTS, analysis._id + "_ALL.zip");
+        // Keep only alphanumeric characters and underscores from user-specified analysis name.
+        String friendlyAnalysisName = filenameCleanString(analysis.name);
+        if (!fileStorage.exists(zippedResultsKey)) {
+            List<FileStorageKey> fileStorageKeys = new ArrayList<>();
+            if (analysis.cutoffsMinutes == null || analysis.travelTimePercentiles == null || analysis.destinationPointSetIds == null) {
+                throw AnalysisServerException.badRequest("Batch result download is not available for legacy regional results.");
+            }
+            // Map from cryptic UUIDs to human readable strings that can replace them.
+            Map<String, String> friendlyReplacements = new HashMap<>();
+            {
+                // Replace the regional analysis ID with its name.
+                friendlyReplacements.put(analysis._id, friendlyAnalysisName);
+                // Replace each destination point set ID with its name.
+                int d = 0;
+                for (String destinationPointSetId : analysis.destinationPointSetIds) {
+                    OpportunityDataset opportunityDataset = Persistence.opportunityDatasets.findByIdIfPermitted(
+                            destinationPointSetId,
+                            userPermissions
+                    );
+                    checkNotNull(opportunityDataset, "Opportunity dataset could not be found in database.");
+                    // Avoid name collisions, prepend an integer.
+                    String friendlyDestinationName = "D%d_%s".formatted(d++, filenameCleanString(opportunityDataset.name));
+                    friendlyReplacements.put(destinationPointSetId, friendlyDestinationName);
```

Comment:
The destination name is used in this new method here, and (part of the) UUID is used in the revised `getRegionalResults` below.

Would a unified method, with the last six characters of the UUID appended to the name, be useful in both places?

Reply:
I think the reason the two use different naming schemes is that including the user-specified name of the destination set means cascading N database lookups to find those names. I was willing to introduce this potential source of lag or bugs in the multi-grid case. It needs to zip things and move files around so might be expected to lag a little. The multi-grid case was also new and intended to be manually triggered.

On the other hand, I was trying to keep the existing single-cutoff codepath more stable, so didn't want to cascade a new database operation there.

In the multi-grid case, I don't really like this UUID-to-name string substitution approach. Practically speaking it works fine. The principle behind it is even reasonable (a one way mapping from totally unique ID strings are equivalent to specific sanitized human-readable strings). But it feels like a workaround, avoiding producing the human readable names in the loop and associating them with the keys.

Assuming we're willing to cascade the database fetch in the single-grid case, what might make sense is to factor out not only a function that generates the short dataset name, but one that creates the whole human-readable name of a single-cutoff grid, then reuse both of these functions in both the single and multi-grid use cases. There's a slight complication in how the extension of the human-readable filename is kept separate in case the raw human readable name is truncated. Maybe this calls for a new data type combining or associating a FileStorageKey with a separate human-readable name and extension, or that truncates and combines the raw human readable name and extension in its constructor.